### PR TITLE
Fix integrating external static frameworks with resource bundles

### LINKED ISSUE #6541: GoogleMaps SDK Bundle can't be found when linking against framework
### What happened?

- We'd like to start using the recently introduced SPM support for the [Google Maps SDK (version 8.4.0)](https://github.com/googlemaps/ios-maps-sdk/tree/8.4.0)
- The Google Maps SDK is **added as a dependency to a framework target** instead of the app target
- The bundle fails to load when initialising the SDK

### How do we reproduce it?

**Reproduction steps**
1. Open the attached sample project: [tuist-googlemaps.zip](https://github.com/user-attachments/files/16347317/tuist-googlemaps.zip)
2. Run `mise x -- tuist install && mise x -- tuist generate`
3. Run the project from Xcode and observe the crash when initialising a `GMSMapView`

**Workaround**
1. Adding an extra copy step to the framework configuration bypasses the issue but duplicates the bundle (and it's very brittle as it references the `.build` folder)

```
.glob(pattern: .relativeToRoot(".build/checkouts/**/GoogleMaps.bundle"))
```

### Error log

```
WARNING: Unable to find and load 'GoogleMaps.bundle' for Google Maps SDK for iOS. This may be a sign that you've forgotten to include a resources bundle in your 'Copy Bundle Resources' build phase. As this bundle contains important resources, you may encounter missing images, translations and other incorrect behavior.
*** Terminating app due to uncaught exception 'GMSServicesException', reason: 'Google Maps SDK for iOS requires GoogleMaps.bundle to be part of your target under 'Copy Bundle Resources''
```

### macOS version

14.5

### Tuist version

4.21.2

### Xcode version

15.1.0

### PR DESCRIPTION
**Pull request by fortmarek**
Resolves https://github.com/tuist/tuist/issues/6541

### Short description 📝

The graph in the issue boils down:
- App -> Dynamic framework -> Static framework -> Resource bundle

In the latest `tuist` version, running the app in such a configuration leads to the following `.app` contents:
```
- App.app
  --> App (executable)
  --> Frameworks
      --> DynamicFramework.framework
          --> StaticFramework_StaticFramework.bundle
```

Correspondingly, looking at the `Copy Resource Bundles` build phase, the `StaticFramework_StaticFramework.bundle` is present only in the `DynamicFramework`. This is expected and it's correct. The issue happens in two scenarios:
- `DynamicFramework` doesn't access the bundle resource accessor, only `App` does
- Some dependencies, like GoogleMaps, depend on implicit SPM behavior of putting the bundle in top level directory (`App.app`)

In the former case, what happens is that the `BundleFinder` symbol is actually included in the `App` target, not the `DynamicFramework`. When trying to find the resource bundle, it tries to look it in the `App.app` directory instead of `App.app/Frameworks/DynamicFramework.framework`. There are two possible solutions for this:
- The `BundleFinder` needs to be access in `DynamicFramework`, forcing the inclusion of the symbols in `DynamicFramework`, not `App`
- Set `GENERATE_MASTER_OBJECT_FILE`to `YES` for `StaticFramework` -> that way, _all_ symbols of `StaticFramework` are included in `DynamicFramework`, not `App`

Setting the `GENERATE_MASTER_OBJECT_FILE` to `YES` when we detect such a scenario is feasible, but it does mean possibly increasing the final bundle size.

Not only that, it doesn't even fix the GoogleMaps scenario because GoogleMaps SDK depends on the implicit behavior of SPM where SPM includes the resource bundle always in the `.app` and not in the first target that can host resources of the static framework 😞 

To make sure dependencies like GoogleMaps can be integrated, we need to copy the behavior. I updated the logic, so that dynamic framework is not considered as a candidate for hosting resources of a static _external_ framework – only products that can embed other products, such as `.app`, are considered in this scenario.

### How to test the changes locally 🧐

Generate `app_with_google_maps` -> the app should no longer crash in runtime.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistCore/Graph/GraphTraverser.swift
Comment (fortmarek) [inline Sources/TuistCore/Graph/GraphTraverser.swift:1]: This file contains the key changes to fix the issue

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 1]
```diff
@@ -167,6 +167,7 @@ public class GraphTraverser: GraphTraversing {
     }
 
     public func resourceBundleDependencies(path: Path.AbsolutePath, name: String) -> Set<GraphDependencyReference> {
+        print(graph.dependencies)
```
Comment (pepicrft) [inline Sources/TuistCore/Graph/GraphTraverser.swift:170]: Let's remove this one.

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 2]
```diff
@@ -176,11 +177,24 @@ public class GraphTraverser: GraphTraversing {
 
         let bundles = filterDependencies(
             from: .target(name: name, path: path),
-            test: isDependencyResourceBundle,
+            test: { dependency in
+                isDependencyResourceBundle(dependency: dependency) && !isDependencyExternal(dependency)
+            },
             skip: canHostResources
         )
+        let externalBundles = filterDependencies(
```
Comment (pepicrft) [inline Sources/TuistCore/Graph/GraphTraverser.swift:185]: What about adding a documentation comment here? Alternatively, they can use the Git history to gain the context, but I thought saving that indirection would be useful when looking at this code in the future.

**Review by pepicrft — APPROVED** [review]
Brilliant ⭐

File: Sources/TuistCore/Graph/GraphTraverser.swift (hunk 1 above)
Comment (fortmarek) [reply Sources/TuistCore/Graph/GraphTraverser.swift]: oops

File: Sources/TuistCore/Graph/GraphTraverser.swift (hunk 2 above)
Comment (fortmarek) [reply Sources/TuistCore/Graph/GraphTraverser.swift:187]: Agree, added 👍

**Comment by Ernest0-Production** [comment]
Apologies if my question is stupid: does this fix other old issue https://github.com/tuist/tuist/issues/5761 related to resource bundle in static xcframework?

**Comment by bvirlet** [comment]
> To make sure dependencies like GoogleMaps can be integrated, we need to copy the behavior. I updated the logic, so that dynamic framework is not considered as a candidate for hosting resources of a static external framework – only products that can embed other products, such as .app, are considered in this scenario.

I know it's a very old issue @fortmarek, but this causes a problem like in https://github.com/tuist/tuist/issues/7126 when using Tuist to build a _SDK_ (so a dynamic framework) as opposed to an app. Since we have to update to the latest Tuist to be unblocked with Xcode 27, we'll add a workaround with a build script to copy the bundles into the dynamic framework.

**Pull request merged**